specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts

```diff
@@ -0,0 +1,342 @@
+import { f } from '../../utils/functools';
+import { overwriteReadOnly, RA } from '../../utils/types';
+import { AnySchema, TableFields } from './helperTypes';
+import {
+  checkPrepAvailability,
+  getTotalLoaned,
+  getTotalResolved,
+  interactionCache,
+  updateLoanPrep,
+} from './interactionBusinessRules';
+import { SpecifyResource } from './legacyTypes';
+import { BusinessRuleResult } from './businessRules';
+import { schema } from './schema';
+import { Collection } from './specifyModel';
+import {
+  BorrowMaterial,
+  CollectionObject,
+  Determination,
+  DNASequence,
+  GiftPreparation,
+  LoanPreparation,
+  LoanReturnPreparation,
+  Tables,
+  Taxon,
+} from './types';
+import * as uniquenessRules from './uniquness_rules.json';
+
+export type BusinessRuleDefs<SCHEMA extends AnySchema> = {
+  readonly onRemoved?: (
+    resource: SpecifyResource<SCHEMA>,
+    collection: Collection<SCHEMA>
+  ) => void;
+  readonly uniqueIn?: UniquenessRule<SCHEMA>;
+  readonly customInit?: (resource: SpecifyResource<SCHEMA>) => void;
+  readonly fieldChecks?: {
+    [FIELDNAME in TableFields<SCHEMA> as Lowercase<FIELDNAME>]?: (
+      resource: SpecifyResource<SCHEMA>
+    ) => Promise<BusinessRuleResult | undefined> | void;
+  };
+};
+
+export const uniqueRules: UniquenessRules = uniquenessRules;
+
+type UniquenessRules = {
+  [TABLE in keyof Tables]?: UniquenessRule<Tables[TABLE]>;
+};
+
+export type UniquenessRule<SCHEMA extends AnySchema> = {
+  [FIELDNAME in TableFields<SCHEMA> as Lowercase<FIELDNAME>]?:
```

Comment:
The reason I use Lowercase<FIELDNAME> is because Backbone has all of the fields as Lowercase, and is not case-insensitive. 
If we use the camel case versions of the field names, Backbone will not be able to recognize a lot of fields and thus the field checks will not be run. 

As a workaround, I suppose I can convert the field names to lowercase as soon as they are passed into `businessRules.ts`, that way functionality remains _and_ we get more readability where it matters

Reply:
uhh, backbone strikes again
yeah, a temporary workaround sounds like a good idea, given that we are getting rid of backbone

Reply:
Fixed in c18fd80 and 7252156